src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,183 @@
             _inRoot = parseResult.GetValue(SlnAddParser.InRootOption);
-            string relativeRoot = parseResult.GetValue(SlnAddParser.SolutionFolderOption);
-
-            SlnArgumentValidator.ParseAndValidateArguments(_fileOrDirectory, _arguments, SlnArgumentValidator.CommandType.Add, _inRoot, relativeRoot);
-
-            bool hasRelativeRoot = !string.IsNullOrEmpty(relativeRoot);
-
-            if (hasRelativeRoot)
-            {
-                relativeRoot = PathUtility.GetPathWithDirectorySeparator(relativeRoot);
-                _relativeRootSolutionFolders = relativeRoot.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries);
-            }
-            else
-            {
-                _relativeRootSolutionFolders = null;
-            }
+            _solutionFolderPath = parseResult.GetValue(SlnAddParser.SolutionFolderOption);
+            SlnArgumentValidator.ParseAndValidateArguments(_fileOrDirectory, _projects, SlnArgumentValidator.CommandType.Add, _inRoot, _solutionFolderPath);
         }
 
         public override int Execute()
         {
-            SlnFile slnFile = SlnFileFactory.CreateFromFileOrDirectory(_fileOrDirectory);
-
-            var arguments = (_parseResult.GetValue<IEnumerable<string>>(SlnAddParser.ProjectPathArgument) ?? Array.Empty<string>()).ToList().AsReadOnly();
-            if (arguments.Count == 0)
+            string solutionFileFullPath = SlnCommandParser.GetSlnFileFullPath(_fileOrDirectory);
+            if (_projects.Count == 0)
             {
                 throw new GracefulException(CommonLocalizableStrings.SpecifyAtLeastOneProjectToAdd);
             }
-
-            PathUtility.EnsureAllPathsExist(arguments, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
-
-            var fullProjectPaths = _arguments.Select(p =>
+            try
             {
-                var fullPath = Path.GetFullPath(p);
-                return Directory.Exists(fullPath) ?
-                    MsbuildProject.GetProjectFileFromDirectory(fullPath).FullName :
-                    fullPath;
-            }).ToList();
-
-            var preAddProjectCount = slnFile.Projects.Count;
-
-            foreach (var fullProjectPath in fullProjectPaths)
+                PathUtility.EnsureAllPathsExist(_projects, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
+                IEnumerable<string> fullProjectPaths = _projects.Select(project =>
+                {
+                    var fullPath = Path.GetFullPath(project);
+                    return Directory.Exists(fullPath) ? MsbuildProject.GetProjectFileFromDirectory(fullPath).FullName : fullPath;
+                });
+                AddProjectsToSolutionAsync(solutionFileFullPath, fullProjectPaths, CancellationToken.None).Wait();
+                return 0;
+            }
+            catch (GracefulException)
```

Comment:
The other catch explicitly excludes this case, and this case just throws, so we can skip this block.